api/src/middlewares/concurrency-lock.ts

```diff
@@ -0,0 +1,58 @@
+import { NextFunction, Response } from 'express';
+import { AuthenticatedRequest } from '../models/types/verification';
+import { getLock, insertLock, removeLock } from '../database/concurrency-lock';
+import { StatusCodes } from 'http-status-codes';
+import { Logger } from '../utils/logger/index';
+import _ from 'lodash';
+
+export enum ConcurrecnyLocks {
+	ChannelLock = 'channel-lock',
+	CredentialLock = 'credential-lock'
+}
+
+const concurrencyLock = (lockName: string) => async (req: AuthenticatedRequest, res: Response, next: NextFunction) => {
+	const afterResponse = async () => {
+		res.removeListener('finish', afterResponse);
+		await releaseConcurrencyLock(lockName)(); // release the lock after next() is finished
+	};
+	res.on('finish', afterResponse);
```

Comment:
It would be better to implement also res.on("close") (https://ipirozhenko.com/blog/measuring-requests-duration-nodejs-express/) because it covers also endpoints that close connection before  getting the response.